File: typescript/src/flatMap.ts
import {listService} from "./listService";
import {failure, Result, success} from "for-comprehension-ts";

const flatMapped: Result<number[], string> =
    success<number, string>(4)
    .flatMap(list => listService(list))
    .flatMap(list => list.length == 0 ? failure("Empty list")
                                      : success(list));

console.log(flatMapped)

File: swift/FlatMap.swift

func flatMap() {
    printName()
    
    let flatMapped: Result<[Int], String> = Result.success(4)
        .flatMap(listService)
        .flatMap { list in
            (list.isEmpty) ? .failure("Empty list") : .success(list)
        }

    print(flatMapped)
}
